packages/core/src/stateUtils.ts

```diff
@@ -259,15 +259,49 @@ export const isStateId = (str: string) => str[0] === STATE_IDENTIFIER;
 
 export function getCandidates<TEvent extends EventObject>(
   stateNode: StateNode<any, TEvent>,
-  eventName: TEvent['type'] | NullEvent['type'] | '*'
+  eventName: TEvent['type'] | NullEvent['type'],
+  /**
+   * If `true`, will use SCXML event token matching semantics
+   */
+  matchTokens: boolean = false
 ): Array<TransitionDefinition<any, TEvent>> {
   const transient = eventName === NULL_EVENT;
   const candidates = stateNode.transitions.filter((transition) => {
-    const sameEventType = transition.eventType === eventName;
-    // null events should only match against eventless transitions
-    return transient
-      ? sameEventType
-      : sameEventType || transition.eventType === WILDCARD;
+    // First, check the trivial case: event names are exactly equal
+    if (transition.eventType === eventName) {
+      return true;
+    }
+
+    // Then, check if transition is a wildcard transition,
+    // which matches all non-transient events
+    if (transition.eventType === WILDCARD) {
+      return !transient;
+    }
+
+    if (!matchTokens && !transition.eventType.endsWith('.*')) {
+      return false;
+    }
+
+    const partialEventTokens = transition.eventType.split('.');
+    const eventTokens = eventName.split('.');
+
+    // tslint:disable-next-line: forin
+    for (const tokenIndex in partialEventTokens) {
+      const partialEventToken = partialEventTokens[tokenIndex];
+      const eventToken = eventTokens[tokenIndex];
+
+      if (partialEventToken === '*') {
+        return true;
+      }
+
+      if (partialEventToken !== eventToken) {
+        return false;
+      }
+
+      continue;
+    }
+
+    return true;
```

Comment:
Please recheck what I'm proposing here because I haven't tested this but I think that this whole chunk of code could be replaced with:
```suggestion
    const matcher = new RegExp(`^${escapeDots(transition.eventType.replace(/\.*/, ''))}(\\.|$)`)
    return matcher.test(eventName)
```
where `escapeDots` would be defined as:
```js
const escapeDots = (str: string): string => str.replace(/\./g, '\\.')
```

--- 

After actually writing this I'm not super sure if this is that much better - but still takes a little bit less of a focus (for me) to understand how this matching works. It's subjective though 🤷‍♂️

Reply:
Actually, according to the SCXML, event names (and descriptors) can only contain alphanumeric characters delimited with dots - not mentioning this as the argument for using regexes here (as I've mentioned - in the end this didn't turn out as clean as I have wanted) but rather to mention that your example is not valid (according to the SCXML).

The only thing I would change in that tokens loop would be to actually iterate over `partialEventToken.length` to obtain indices - you wouldn't have to tell TSLint to shut up then and it would kinda be more idiomatic. Knowing that those are arrays I've assumed that this is for-of loop and was scratching my head how the hell you got access to the current index this way 😅

Reply:
Makes sense - I switched this to a simple `for` loop with indexes.